### PR TITLE
[GH-429] Dungeon Settlement Levels

### PR DESCRIPTION
**Pull request by lotuuu**
## Motivation

Adds DungeonSettlementLevels to Users
Closes https://github.com/lambdaclass/champions_of_mirra/issues/429

## Summary of changes

- Creates DungeonSettlementLevel schema
- Removes kaline_tree_id requirement in user creation
- Creates message for claiming Dungeon Settlement AFK Rewards
  - Did not create a message for _showing_ these rewards because we will be automatically claiming them everytime we open the Dungeon Settlement screen.
- Rename most Kalline Tree AFK Rewards messages and funcs because we now have two different AFK Reward groups
- Add Tests

## How to test it?

You can create a user and check that his dungeon_settlement_level is instantiated at level 1.

If you give your user 50 "Blueprints" (using `GameBackend.Users.Currencies.add_currency_by_name!/3`), you will be able to level up the dungeon settlement of your user with `Champions.Users.level_up_dungeon_settlement/1`. If you get your user again (`Champions.Users.get_user_by_username/1`) you will be able to see that his dungeon_settlement_level has increased, and his "Gold" and "Blueprints" currencies decreased!

Since this implements protobuf messages, this PR also adds SocketTester messages! You can also try those :P

### CODE REVIEW

File: apps/gateway/test/champions_test.exs [hunk 1]
```diff
@@ -689,7 +689,7 @@ defmodule Gateway.Test.Champions do
     end
 
     test "leveling up the Kaline Tree increments the afk rewards", %{socket_tester: socket_tester} do
-      {:ok, user} = Users.register("AfkRewardsUser")
+      {:ok, user} = Users.register("KelineTreeAFKRewardsUser")
```
Comment (ncontinanza) [inline apps/gateway/test/champions_test.exs:692]:
```suggestion
      {:ok, user} = Users.register("KalineTreeAFKRewardsUser")
```

File: apps/serialization/gateway.proto [hunk 2]
```diff
@@ -178,6 +189,16 @@ syntax = "proto3";
     repeated AfkRewardRate afk_reward_rates = 6;
   }
 
+  message DungeonSettlementLevel {
+    string id = 1;
```
Comment (ncontinanza) [inline apps/serialization/gateway.proto:193]: This is unnecessary for the client (the KalineTreeLevel id as well)

**Review by ncontinanza — CHANGES_REQUESTED** [review]

File: apps/champions/lib/champions/users.ex [hunk 3]
```diff
@@ -212,9 +228,17 @@ defmodule Champions.Users do
...
-    afk_rewards = get_afk_rewards(user_id)
+  def claim_kaline_afk_rewards(user_id) do
+    afk_rewards = get_kaline_afk_rewards(user_id)
+    claim_afk_rewards(user_id, afk_rewards, :kaline)
+  end
 
+  def claim_dungeon_afk_rewards(user_id) do
+    afk_rewards = get_dungeon_afk_rewards(user_id)
+    claim_afk_rewards(user_id, afk_rewards, :dungeon)
+  end
```
[the remark this answers is not in the document]
Comment (lotuuu) [reply apps/champions/lib/champions/users.ex:239]: Done!

File: apps/champions/lib/champions/users.ex [hunk 4]
```diff
@@ -200,8 +202,22 @@ defmodule Champions.Users do
...
+          amount = calculate_afk_rewards(user.last_dungeon_afk_reward_claim, reward_rate)
+          %{currency: currency, amount: amount}
+        end)
+
+      {:error, :not_found} ->
+        {:error, :not_found}
+    end
+  end
+
+  defp calculate_afk_rewards(last_claim, afk_reward_rate) do
```
[the remark this answers is not in the document]
Comment (lotuuu) [reply apps/champions/lib/champions/users.ex:220]: Done!

File: apps/champions/lib/champions/users.ex [hunk 5]
```diff
@@ -252,9 +276,23 @@ defmodule Champions.Users do
...
+    with {:user, {:ok, user}} <- {:user, Users.get_user(user_id)},
+         level_up_costs = user.dungeon_settlement_level.level_up_costs,
+         {:can_afford, true} <- {:can_afford, Currencies.can_afford(user_id, level_up_costs)} do
+      Users.level_up_dungeon_settlement(user_id, level_up_costs)
+    else
+      {:can_afford, false} -> {:error, :cant_afford}
+      {:user, {:error, :not_found}} -> {:error, :user_not_found}
+    end
+  end
+
```
[the remark this answers is not in the document]
Comment (lotuuu) [reply apps/champions/lib/champions/users.ex:277]: This abstraction would require us to replace the Kaline Tree level up costs for currency_costs, which is a task that's outside the scope of this PR. It's also a beta feature. I'll create a ticket for it!

**Review by ncontinanza — APPROVED** [review]
LGTM!

File: apps/game_backend/lib/game_backend/users.ex [hunk 6]
```diff
@@ -233,11 +260,43 @@ defmodule GameBackend.Users do
     end
   end
 
+  @doc """
+  Level up the Dungeon Settlement of a user.
+
+  Returns the updated user if the operation was successful.
+  """
+  def level_up_dungeon_settlement(user_id, level_up_costs) do
```
Comment (agustinesco) [inline apps/game_backend/lib/game_backend/users.ex:268]: Isn't this the same as `level_up_kaline_tree`? just changing the `increment_settlement_level`right?

**Review by agustinesco — APPROVED** [review]
LGTM! tested and the currencies can be claimed without issues 🪨

**Pull request merged**